rusty-machine/src/linalg/matrix/iter.rs

```diff
@@ -0,0 +1,194 @@
+use std::marker::PhantomData;
+use std::slice;
+
+use super::{Matrix, MatrixSlice, MatrixSliceMut};
+
+#[derive(Debug)]
+pub struct Rows<'a, T: 'a> {
+    row_start: *const T,
+    row_pos: usize,
+    slice_rows: usize,
+    slice_cols: usize,
+    row_stride: isize,
+    _marker: PhantomData<&'a T>,
+}
+
+/// Iterates over the matrix data by slice
+impl<'a, T> Iterator for Rows<'a, T> {
+    type Item = &'a [T];
```

Comment:
`MatrixSlice<T>` may be a good idea actually - I hadn't really considered it before. I wouldn't use a `Vector<T>` as this needs to own it's data (it uses a `Vec`).

The reason I wrote it right now to return slices is that I want to use this iterator to provide more efficient OpAssign methods for `MatrixSliceMut`. This will be helpful/needed for the multithreaded matrix multiplication work.